modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,439 @@
+			$instance['title'] = $this->defaults['title'];
+
+		if ( ! empty( $instance['title'] ) )
+			echo $args['before_title'] . esc_html( $instance['title'] ) . $args['after_title'];
+
+		if ( ! $instance['token_id'] ) {
+			echo 'This widget cannot make new connections to Instagram. You can install and use a third-party Instagram plugin instead. Please <a href="https://wordpress.com/help/contact/">contact us if you need help</a> setting this up.';
+			echo '<p><em>This notice is only shown to you.</em></p>';
+			echo '<a class="button-primary" target="_top" href="https://public-api.wordpress.com/connect/?action=request&amp;kr_nonce=25051df3f2&amp;nonce=84e239e897&amp;for=instagram-widget&amp;service=instagram&amp;blog=105567107&amp;kr_blog_nonce=217ce9172b&amp;magic=keyring&amp;instagram_widget_id=8">Authorize Instagram Access</a>';
+		} else {
+			$images = $this->get_images( $instance );
+
+			if ( ! is_array( $images ) ) {
+				echo '<p>' . __( 'There was an error retrieving images from Instagram. An attempt will be remade in a few minutes.', 'wpcomsh' ) . '</p>';
+			}
+			elseif ( ! $images ) {
+				echo '<p>' . __( 'No Instagram images were found.', 'wpcomsh' ) . '</p>';
+			} else {
+
+				echo '<div class="' . esc_attr( 'wpcom-instagram-images wpcom-instagram-columns-' . (int) $instance['columns'] ) . '">' . "\n";
+				foreach ( $images as $image ) {
+					echo '<a href="' . esc_url( $image['link'] ) . '" target="' . esc_attr( apply_filters( 'wpcom_instagram_widget_target', '_self' ) ) . '"><div class="sq-bg-image" style="background-image: url(' . esc_url( set_url_scheme( $image['url'] ) ) . ')"><span class="screen-reader-text">' . esc_attr( $image['title'] ) . '</span></div></a>' . "\n";
+				}
+				echo "</div>\n";
+			}
+		}
+
+		echo $args['after_widget'];
+	}
+
+	/**
+	 * Outputs the widget configuration form for the widget administration page.
+	 * Allows the user to add new Instagram Keyring tokens and more.
+	 *
+	 * @param array $instance The widget instance (configuration options).
+	 */
+	public function form( $instance ) {
+		$instance = wp_parse_args( $instance, $this->defaults );
+
+		// If coming back to the widgets page from an action, expand this widget
+		if ( isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number ) {
+			echo '<script type="text/javascript">jQuery(document).ready(function($){ $(\'.widget[id$="wpcom_instagram_widget-' . esc_js( $this->number ) . '"] .widget-inside\').slideDown(\'fast\'); });</script>';
+		}
+
+		// If coming back from an OAuth authentication, validate and use the one in the URL
+		if ( isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number
+			&& ! empty( $_GET['instagram_widget'] ) && 'connection_verified' == $_GET['instagram_widget']
+			&& ! empty( $_GET['token_id'] ) && $instance['token_id'] !== (int) $_GET['token_id'] ) {
+			//$token_store = Keyring::init()->get_token_store();
```

Comment:
These lines are commented out because we don't have Keyring in Jetpack, so we can't verify the token.